### PR TITLE
Add 'union' mode for schema inference

### LINKED ISSUE #55428: Allow smart "merging" of schema for multi-file schema inference
**Use case**

When performing schema inference on multiple files, there is an opportunity to be more flexible by having clickhouse "union_by_name" the schemata to allow these schema variants to load as a unified table. This would be similar to read_parquet('*.parquet', union_by_name=True) to allow schema inference to simply add new columns when they are not present in all files? 

The same logic could apply to CSV, TSV, jsonlines, etc., but my use case is for Parquet, so feel free to scope as you see fit.

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add 'union' mode for schema inference. In this mode the resulting table schema is the union of all files schemas (so schema is inferred from each file). The mode of schema inference is controlled by a setting `schema_inference_mode` with 2 possible values - `default` and `union`. Closes https://github.com/ClickHouse/ClickHouse/issues/55428

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by Avogar** [comment]
TODO: add docs

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 6f61ccfe289d25bbb8594d03a30150baac308f09 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55892/6f61ccfe289d25bbb8594d03a30150baac308f09/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>❌ failure</td></tr>
<tbody>
</table>

File: src/Core/SettingsChangesHistory.h [hunk 1]
```diff
@@ -80,6 +80,9 @@ namespace SettingsChangesHistory
 /// It's used to implement `compatibility` setting (see https://github.com/ClickHouse/ClickHouse/issues/35972)
 static std::map<ClickHouseVersion, SettingsChangesHistory::SettingsChanges> settings_changes_history =
 {
+    {"23.10", {{"input_format_parquet_allow_missing_columns", false, true, "Allow missing columns in Parquet files by default"},
```
Comment (Avogar) [inline src/Core/SettingsChangesHistory.h:83]:
```suggestion
    {"23.11", {{"input_format_parquet_allow_missing_columns", false, true, "Allow missing columns in Parquet files by default"},
```

File: src/Formats/SchemaInferenceUtils.cpp [hunk 2]
```diff
@@ -547,6 +547,54 @@ namespace
...
+                }
+            }
+        }
+
+        /// Try to find common type for each tuple element with the same name.
+        DataTypes element_types;
+        element_types.reserve(names_to_types.size());
+        for (const auto & name : element_names)
+        {
+            auto types = names_to_types[name];
```
Comment (serxa) [inline src/Formats/SchemaInferenceUtils.cpp:580]: Copy is not necessary here
```suggestion
            auto & types = names_to_types[name];
```

**Review by serxa — APPROVED** [review]
Implementations of `IReadBufferIterator` start to look very heavy and a bit repetitive. Maybe we can do something to keep the common code separated? A common base class? But maybe it is okay, at least it is super straightforward.

Anyways, PR looks good to me.

**Comment by serxa** [comment]
`02375_system_schema_inference_cache` failures are related to the changes

**Comment by Avogar** [comment]
> Maybe we can do something to keep the common code separated?

Actually I have a dream to refactor reading from S3/File/HDFS/Azure/URL storages to make some base interfaces, because there are a lot of copying, but don't have time right now :(
But I have it in mind, so hope some day I will do it

**Pull request merged**